Dart/src/com/jetbrains/lang/dart/ide/toolingDaemon/DartToolingDaemonService.kt

```diff
@@ -0,0 +1,161 @@
+import com.intellij.openapi.util.text.StringUtil
+import com.jetbrains.lang.dart.sdk.DartSdk
+import de.roderick.weberknecht.WebSocket
+import de.roderick.weberknecht.WebSocketEventHandler
+import de.roderick.weberknecht.WebSocketException
+import de.roderick.weberknecht.WebSocketMessage
+import java.net.URI
+import java.net.URISyntaxException
+import java.nio.charset.StandardCharsets
+import java.util.concurrent.atomic.AtomicInteger
+
+@Service(Service.Level.PROJECT)
+class DartToolingDaemonService(private val myProject: Project) : Disposable {
+
+  private val MIN_SDK_VERSION: String = "3.4"
+
+  private val STARTUP_MESSAGE_PREFIX = "The Dart Tooling Daemon is listening on "
+
+  private var myWebSocket: WebSocket? = null
+  private val nextRequestId = AtomicInteger()
+
+  private val consumerMap: MutableMap<String, ToolingDaemonConsumer> = Maps.newHashMap()
+
+  private val listeners: MutableList<ToolingDaemonListener> = ArrayList()
+  private lateinit var dtdProcessHandler: ColoredProcessHandler;
+
+  @Throws(ExecutionException::class)
+  fun startService() {
+    val sdk = DartSdk.getDartSdk(myProject)
+    if (sdk == null || !isDartSdkVersionSufficient(sdk)) {
+      return
+    }
+    val result = GeneralCommandLine().withWorkDirectory(sdk.homePath)
+    result.charset = StandardCharsets.UTF_8
+    result.exePath = FileUtil.toSystemDependentName("dart")
+    result.addParameter("tooling-daemon")
+
+    dtdProcessHandler = ColoredProcessHandler(result)
+    dtdProcessHandler.addProcessListener(object : ProcessListener {
+      override fun onTextAvailable(event: ProcessEvent, outputType: Key<*>) {
+        val text = event.text.trim { it <= ' ' }
+        if (text.startsWith(STARTUP_MESSAGE_PREFIX)) {
+          val address: String = text.split(STARTUP_MESSAGE_PREFIX.toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray().get(1)
+          connectToDTDWebsocket(address)
+        }
+      }
+    })
+
+    dtdProcessHandler.startNotify()
+  }
+
+  private fun connectToDTDWebsocket(address: String) {
+    try {
+      myWebSocket = WebSocket(URI("ws://$address/ws"))
+      myWebSocket!!.eventHandler = object : WebSocketEventHandler {
+        override fun onClose() {
+        }
+
+        override fun onMessage(message: WebSocketMessage) {
+          val json: JsonObject
+          try {
+            json = JsonParser.parseString(message.text) as JsonObject
+          }
+          catch (e: Exception) {
+            println("Parse message failed: " + message.text + e)
+            return
+          }
+
+          val method = json["method"]?.asString
+          if (method == "streamNotify") {
+            val params = json["params"].asJsonObject
+            for (listener in java.util.ArrayList(listeners)) {
+              listener.received(params["streamId"].asString, json)
+            }
+          }
+
+          val id = json["id"].asString
+          val consumer = consumerMap.remove(id)
+          consumer?.received(json)
```

Comment:
should this  be in `else` block?

Reply:
Oh yes, you're right. Or return after the for loop ends and listeners have been notified of `streamNotify` events.

Reply:
Ok, pushing the fix